pkg/volume/csi/nodeinfomanager/nodeinfomanager.go

```diff
@@ -456,106 +452,73 @@ func (nim *nodeInfoManager) installDriverToCSINodeInfo(
-		}
-	}
 
 	if !specModified && !statusModified {
 		return nil
 	}
 
 	// Append new driver
-	driverSpec := csiv1alpha1.CSIDriverInfoSpec{
+	driverSpec := storage.CSINodeDriver{
 		Name:         driverName,
 		NodeID:       driverNodeID,
 		TopologyKeys: topologyKeys.List(),
 	}
-	driverStatus := csiv1alpha1.CSIDriverInfoStatus{
-		Name:      driverName,
-		Available: true,
-		// TODO(https://github.com/kubernetes/enhancements/issues/625): Add actual migration status
-		VolumePluginMechanism: csiv1alpha1.VolumePluginMechanismInTree,
-	}
 
 	newDriverSpecs = append(newDriverSpecs, driverSpec)
-	newDriverStatuses = append(newDriverStatuses, driverStatus)
 	nodeInfo.Spec.Drivers = newDriverSpecs
-	nodeInfo.Status.Drivers = newDriverStatuses
 
-	err := validateCSINodeInfo(nodeInfo)
-	if err != nil {
-		return err
-	}
-	_, err = csiKubeClient.CsiV1alpha1().CSINodeInfos().Update(nodeInfo)
+	_, err := csiKubeClient.StorageV1beta1().CSINodes().Update(nodeInfo)
 	return err
 }
 
-func (nim *nodeInfoManager) uninstallDriverFromCSINodeInfo(
+func (nim *nodeInfoManager) uninstallDriverFromCSINode(
 	csiDriverName string) error {
 
-	csiKubeClient := nim.volumeHost.GetCSIClient()
+	csiKubeClient := nim.volumeHost.GetKubeClient()
 	if csiKubeClient == nil {
 		return fmt.Errorf("error getting CSI client")
 	}
 
 	var updateErrs []error
 	err := wait.ExponentialBackoff(updateBackoff, func() (bool, error) {
-		if err := nim.tryUninstallDriverFromCSINodeInfo(csiKubeClient, csiDriverName); err != nil {
+		if err := nim.tryUninstallDriverFromCSINode(csiKubeClient, csiDriverName); err != nil {
 			updateErrs = append(updateErrs, err)
 			return false, nil
 		}
 		return true, nil
 	})
 	if err != nil {
-		return fmt.Errorf("error updating CSINodeInfo: %v; caused by: %v", err, utilerrors.NewAggregate(updateErrs))
+		return fmt.Errorf("error updating CSINode: %v; caused by: %v", err, utilerrors.NewAggregate(updateErrs))
 	}
 	return nil
 }
 
-func (nim *nodeInfoManager) tryUninstallDriverFromCSINodeInfo(
-	csiKubeClient csiclientset.Interface,
+func (nim *nodeInfoManager) tryUninstallDriverFromCSINode(
+	csiKubeClient clientset.Interface,
 	csiDriverName string) error {
 
-	nodeInfoClient := csiKubeClient.CsiV1alpha1().CSINodeInfos()
+	nodeInfoClient := csiKubeClient.StorageV1beta1().CSINodes()
 	nodeInfo, err := nodeInfoClient.Get(string(nim.nodeName), metav1.GetOptions{})
-	if err != nil {
-		return err // do not wrap error
+	if err != nil && errors.IsNotFound(err) {
+		return nil
+	} else if err != nil {
+		return err
 	}
 
-	hasModified := false
-	newDriverStatuses := []csiv1alpha1.CSIDriverInfoStatus{}
-	for _, driverStatus := range nodeInfo.Status.Drivers {
-		if driverStatus.Name == csiDriverName {
-			// Uninstall the driver if we find it
-			hasModified = driverStatus.Available
-			driverStatus.Available = false
+	// Uninstall matching CSINodeDriver
+	for i, driver := range nodeInfo.Spec.Drivers {
+		if driver.Name == csiDriverName {
```

Comment:
I think we want the reverse.  We should not include the driver in the list if it's the one we're trying to unregister.

Reply:
The logic here is actually correct but I think it's a little bit confusing. Would prefer to revert it to the version where we append each driver in the list and skip the driver we're trying to unregister.

(this current method also has very strange behavior in the case the driver is not found in the drivers list which is kind of like a silent default)

Reply:
Ok